### ISSUE TITLE
Runtime exception for Android OS v12+ 

### ISSUE THREAD

**Issue by imkrishh** [infobip/mobile-messaging-react-native-plugin#50]
neact native infobip runtime crash on Android

`infobip-mobile-messaging-react-native-plugin` -> v6.3.3
`react-native` -> v0.65.1
`compileSdkVersion` ->  31
`targetSdkVersion` ->  31

> Fatal Exception: java.lang.IllegalArgumentException: com.bundlex.bundlex: Targeting S+ (version 31 and above) requires that one of FLAG_IMMUTABLE or FLAG_MUTABLE be specified when creating a PendingIntent.
Strongly consider using FLAG_IMMUTABLE, only use FLAG_MUTABLE if some functionality depends on the PendingIntent being mutable, e.g. if it needs to be used with inline replies or bubbles.
       at android.app.PendingIntent.checkFlags(PendingIntent.java:382)
       at android.app.PendingIntent.getActivitiesAsUser(PendingIntent.java:616)
       at android.app.PendingIntent.getActivities(PendingIntent.java:598)
       at android.app.PendingIntent.getActivities(PendingIntent.java:545)
       at org.infobip.mobile.messaging.notification.BaseNotificationHandler.createTapPendingIntent(BaseNotificationHandler.java:230)
       at org.infobip.mobile.messaging.notification.BaseNotificationHandler.createNotificationCompatBuilder(BaseNotificationHandler.java:117)
       at org.infobip.mobile.messaging.interactive.notification.InteractiveNotificationHandler.getNotificationBuilder(InteractiveNotificationHandler.java:57)
       at org.infobip.mobile.messaging.interactive.notification.InteractiveNotificationHandler.displayNotification(InteractiveNotificationHandler.java:44)
       at org.infobip.mobile.messaging.cloud.MobileMessageHandler.handleMessage(MobileMessageHandler.java:72)
       at org.infobip.mobile.messaging.mobileapi.messages.MessagesSynchronizer$1.after(MessagesSynchronizer.java:92)
       at org.infobip.mobile.messaging.mobileapi.messages.MessagesSynchronizer$1.after(MessagesSynchronizer.java:72)
       at org.infobip.mobile.messaging.mobileapi.common.MRetryableTask$1.after(MRetryableTask.java:92)
       at org.infobip.mobile.messaging.mobileapi.common.MAsyncTask$2.onPostExecute(MAsyncTask.java:73)
       at org.infobip.mobile.messaging.mobileapi.common.MAsyncTask$2.onPostExecute(MAsyncTask.java:43)
       at android.os.AsyncTask.finish(AsyncTask.java:771)
       at android.os.AsyncTask.access$900(AsyncTask.java:199)
       at android.os.AsyncTask$InternalHandler.handleMessage(AsyncTask.java:788)
       at android.os.Handler.dispatchMessage(Handler.java:106)
       at android.os.Looper.loopOnce(Looper.java:226)
       at android.os.Looper.loop(Looper.java:313)
       at android.app.ActivityThread.main(ActivityThread.java:8669)
       at java.lang.reflect.Method.invoke(Method.java)
       at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:571)
       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1135)

**Comment by riskpp** [comment]
Hi @imkrishh, could you please describe which notification are you tapping while receiving this crash? Does it contain browserUrl, webViewUrl or Deeplink?
It sounds weird because PendingIntent always contains flag "FLAG_MUTABLE"

**Comment by rohitsurawat** [comment]
@riskpp  this issue is fatal because i am also facing this issue 
org.infobip.mobile.messaging.notification.BaseNotificationHandler.createTapPendingIntent
I have one application which is live and we are getting 95% of customer unable to start the application please help here @riskpp

**Comment by tjuric** [comment]
@rohitsurawat and @imkrishh, please describe your exact use cases which are producing this behaviour. Which payload does notification contain? Do you use any of: browserUrl, webViewUrl or deeplink?

**Comment by rohitsurawat** [comment]
@tjuric  @riskpp 
**Strongly consider using FLAG_IMMUTABLE,** only **use FLAG_MUTABLE if some functionality depends on the PendingIntent being mutable,** e.g. if it needs to be used with inline replies or bubbles.


If you read the stack trace it is clearly stating that while passing the flag Strongly Consider FLAG_IMMUTABLE , only use FLAG_MUTABLE in case if your notification have reply or bubble functionality associated to it.

@tjuric  if you see @riskpp  last reply according to her it will always be  "FLAG_MUTABLE" which is wrong in case of simple notifications.


Also in infobip android sdk 
Class : MobileMessagingProperty have no decleared enum for FLAG_IMMUTABLE .

Check Line no 59 for android sdk

 

> PENDING_INTENT_FLAGS("org.infobip.mobile.messaging.notification.PENDING_INTENT_FLAGS", (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) ? PendingIntent.FLAG_CANCEL_CURRENT | PendingIntent.FLAG_MUTABLE : PendingIntent.FLAG_CANCEL_CURRENT),



Please help us try to resolve this asap we are getting impacted.
Also will share the payload that we are sending asap

**Comment by rohitsurawat** [comment]
@riskpp  @tjuric  any solution here 
We are waiting for your response

**Comment by rohitsurawat** [comment]
@tjuric @riskpp  any update

**Comment by riskpp** [comment]
So I've tested it with "WEB_VIEW_URL" (on the tap webView is opening) on Android 12 emulator (31 API) and I can't reproduce any crashes. If talking about newer Android versions, we don't support them for now, but have tasks for that in our backlog and will work on it soon. We will include the possible update for the flag FLAG_MUTABLE to the FLAG_IMMUTABLE, for the cases when push notification doesn't have possibility to be replied to one of the tasks.

**Comment by riskpp** [comment]
Actually you may try to update react native at least to the 0.68 (I was testing on this) as far as I can see on the documentation, RN started to support 31st Android API from 0.68 version.

**Comment by rohitsurawat** [comment]
@riskpp  We were facing huge volume of crashes due to notifications  generated from Infobip Portal so 4 days ago we decide to stop delivery of notification, Since than we have faced 111 crashes in production. And from our testing we came to conclusion that code crash at the time notification is created in Android 12 or Higher mobile. So user will no experience any thing. But our business is impacted due to this. 

As i can read your older comment you were saying something about change to FLAG_IMMUTABLE , Can you give us an estimated time when we can expect an infobip library update so that we can update test and release a new version for our customers.

**Comment by rohitsurawat** [comment]
@riskpp  We have sent email regarding payload and campaign id please check with infobip support team. However we were not able to extract payload or campaign id from portal please provide us clear guide how to extract this info if possible then please respond via email only

**Comment by rohitsurawat** [comment]
Hi @riskpp  @tjuric @mstipanov 

As suggested by @riskpp  we upgraded the versions.

react native : 0.68.3

Infobip version right now is :
infobip-mobile-messaging-react-native-plugin: ^6.3.4

But still same crash we shared earlier is coming.
Please help us otherwise we are afraid don't have any other option than dropping your service and choose some other service.

**Comment by riskpp** [comment]
Hi @quantumarun, @rohitsurawat,
Sorry for the delay, we've sent an answer by the e-mail, just to have a followup here:

We’ve check the setup with test application similar to yours and can’t reproduce the issue anyhow. Also we’ve checked your campaign by the provided id and found that you probably have installation of the apps, to which you’ve send the push notifications with version of the react-native plugin, which do not support Android 12 and don’t have the mutability flag set, that could lead to the crash.

BR, Olga

**Comment by riskpp** [comment]
@quantumarun , @rohitsurawat ,
Could you please clarify, have you managed to resolve the issue by that advice or it still persists?

**Comment by riskpp** [comment]
We've released actually the change to the FLAG_IMMUTABLE where mutability not required, but still It might be not the reason of why this appears
https://github.com/infobip/mobile-messaging-react-native-plugin/releases/tag/6.3.6

**Comment by imsugat** [comment]
Hi @riskpp,

I am also facing the same issue and have tried the fix suggested in [#issuecomment-1269770779](https://github.com/infobip/mobile-messaging-react-native-plugin/issues/50#issuecomment-1269770779) but it's not working. I am getting this crash in Google Crashlytics so not able to capture the notification payload.

infobip-mobile-messaging-react-native-plugin -> v6.3.5
react-native -> v0.66.5
compileSdkVersion -> 31
targetSdkVersion -> 31

**Comment by riskpp** [comment]
Hi @imsugat are you sure that the user which got the crash updated the version? I can reproduce this crash only if user has Android 12 and version of the plugin, which doesn't support Android 12 ( < 6.0.0).

**Comment by riskpp** [comment]
I've just found one possible edge case when it could happen:
1. User updated the application from the version without Android 12 support to the version with Android 12 support (plugin version > 6.0.0) and never run the app (`mobilemesaging.init` wasn't called).
2. Application received notification when it's is killed
3. `mobilemesaging.init` wasn't called on the background when notification arrived

In this case might happen that property for PendingIntent flags, saved in the Preferences wasn't updated to the new one (which contains FLAG_IMMUTABLE).
We'll try to fix that, but before you from your side could try to be sure that you calling `mobilemessaging.init` if application is started by receiving the push notification.

**Comment by riskpp** [comment]
Version with possible fix is released https://github.com/infobip/mobile-messaging-react-native-plugin/releases/tag/7.0.3

Check the [Migration guide](https://github.com/infobip/mobile-messaging-react-native-plugin/wiki/Migration-guides#migration-from-6x-to-7x) if you updating from 6.x to 7.x

**Comment by imsugat** [comment]
Hi @riskpp I have tried updating the version of the Infobip plugin in Medgami app to 7.0.3 as suggested. While updating I am getting cli error

npm ERR! code ERESOLVE
npm ERR! ERESOLVE unable to resolve dependency tree
npm ERR! 
npm ERR! While resolving: 
npm ERR! Found: react-native@0.66.5
npm ERR! node_modules/react-native
npm ERR!   react-native@"0.66.5" from the root project
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! peer react-native@">=0.68.0 <1.0.x" from infobip-mobile-messaging-react-native-plugin@7.0.3
npm ERR! node_modules/infobip-mobile-messaging-react-native-plugin
npm ERR!   infobip-mobile-messaging-react-native-plugin@"7.0.3" from the root project
npm ERR! 
npm ERR! Fix the upstream dependency conflict, or retry
npm ERR! this command with --force, or --legacy-peer-deps
npm ERR! to accept an incorrect (and potentially broken) dependency resolution.

Same issue I have shared via mail.
